File: packages/client-java/src/main/java/com/durablestreams/exception/DurableStreamException.java
package com.durablestreams.exception;

import java.util.Map;
import java.util.Optional;

/**
 * Base exception for all Durable Streams errors.
 */
public class DurableStreamException extends RuntimeException {
    private final Integer statusCode;
    private final String errorCode;
    private final Map<String, String> responseHeaders;

    public DurableStreamException(String message) {
        this(message, null, null, null, null);
    }

    public DurableStreamException(String message, Throwable cause) {
        this(message, cause, null, null, null);
    }

    public DurableStreamException(String message, Integer statusCode) {
        this(message, null, statusCode, null, null);
    }

    public DurableStreamException(String message, Integer statusCode, String errorCode) {
        this(message, null, statusCode, errorCode, null);
    }

    public DurableStreamException(String message, Throwable cause, Integer statusCode,
                                   String errorCode, Map<String, String> responseHeaders) {
        super(message, cause);
        this.statusCode = statusCode;
        this.errorCode = errorCode;
        this.responseHeaders = responseHeaders;
    }

    public Optional<Integer> getStatusCode() {
        return Optional.ofNullable(statusCode);
    }

    public Optional<String> getErrorCode() {
        return Optional.ofNullable(errorCode);
    }

    public Map<String, String> getResponseHeaders() {
        return responseHeaders != null ? responseHeaders : Map.of();
    }
}

File: packages/durable-streams/packages/client-php/src/Exception/DurableStreamException.php
<?php

declare(strict_types=1);

namespace DurableStreams\Exception;

/**
 * Base exception for all Durable Streams errors.
 */
class DurableStreamException extends \Exception
{
    public function __construct(
        string $message,
        protected ?string $errorCode = null,
        protected ?int $httpStatus = null,
        protected array $headers = [],
        ?\Throwable $previous = null,
    ) {
        parent::__construct($message, 0, $previous);
    }

    public function getErrorCode(): ?string
    {
        return $this->errorCode;
    }

    public function getHttpStatus(): ?int
    {
        return $this->httpStatus;
    }

    public function getHeaders(): array
    {
        return $this->headers;
    }

    public function isRetryable(): bool
    {
        return in_array($this->httpStatus, [429, 500, 502, 503, 504], true);
    }
}
